file.go

```diff
@@ -714,6 +718,22 @@ func getBufioReader(r io.Reader) *bufio.Reader {
 func putBufioReader(rbuf *bufio.Reader) {
 	if rbuf != nil {
 		rbuf.Reset(nil)
-		bufioReaderPool.Put(rbuf)
+		getBufioReaderPool(rbuf.Size()).Put(rbuf)
```

Comment:
There are cases where the `bufio` package does not retain the input size, which would case the buffers to be released to a different pool than the one they were created from; for example https://cs.opensource.google/go/go/+/refs/tags/go1.19:src/bufio/bufio.go;l=53-57

There are two ways we could address: either pass the buffer size as input parameter, or return both the `*bufio.Reader` and `*sync.Pool` from `getBufioReader` so the caller has access to the pool to release the buffer to. The latter has the advantage of not requiring a second synchronization and lookup of the pool when releasing the reader.

Reply:
I reviewed the `bufio.Reader` logic and, as far as I can tell, it doesn't resize itself during normal operation.

Although, you are right that if you create a reader with a buffer size of < 16 it will increase the size to 16. I suppose this would mean that if someone were to open up different parquet files in the same process with very small buffer sizes they'd all be pooled together with size 16?

As is I don't think the `bufio.Reader` behavior is concerning, although who knows what future versions of Go will bring. We could do something like drop a reader on put whose size doesn't exist in the map? or panic?

Reply:
> Although, you are right that if you create a reader with a buffer size of < 16 it will increase the size to 16. I suppose this would mean that if someone were to open up different parquet files in the same process with very small buffer sizes they'd all be pooled together with size 16?

I believe the consequence would be that no pooling would apply, parquet-go would always look for buffers in the pool matching the buffer size exactly, then release them to the buffer pool for size 16.

I agree it's a corner case, but it highlights that the `bufio.Reader` API does not promise to respect the buffer size that it is constructed with; robust software must not rely on it as we cannot anticipate the consequences if the standard library implementation changes (which it could rightfully do).

> We could do something like drop a reader on put whose size doesn't exist in the map? or panic?

Do you have concerns about the strategy I described above?
```go
rbuf, pool := getBufioReader(bufferSize)
defer pool.Put(rbuf)
```

Reply:
Pushed a commit that returns the pool from `getBufioReader()` along with the buffer. No real concerns. It did require adding another struct  field on `filePages` but that seems minor.

I'm still using the `putBufioReader()` method to place it back in the pool so that the buffer is correctly reset. I suppose I could return a `put()` function from `getBufioReader()` instead? I have no strong feelings either way.

Reply:
Returning a function may cause the compiler to add an extra heap allocation to capture the `*sync.Pool` value to release the object to.

I think the current form is robust and clear, let's go with it 👍